### ISSUE TITLE
CloseViewModelAsync does not call UninitializeModelInternal

### ISSUE THREAD

**Issue by lipchev** [Catel/Catel#1716]
## Please check all of the platforms you are having the issue on (if platform is not listed, it is not supported)

 - [x] WPF
 - [ ] UWP
 - [ ] iOS
 - [ ] Android
 - [ ] .NET Standard
 - [ ] .NET Core

## Component

ViewModelBase

What component is this issue occurring in?

## Version of Library

5.12.11-beta0009 

## Version of OS(s) listed above with issue

win 10

## Steps to Reproduce
1. Create a ViewModel with a Model property
2. `CloseViewModelAsync` the view model without calling either `SaveViewModelAsync` or `CancelViewModelAsync`
3. Observe: a memory leak is created for the view model instance (not cleaning up the ModelErrorInfo subscriptions)

## Expected Behavior

CloseViewModelAsync should (also) perform the model object cleanup 

## Actual Behavior

![image](https://user-images.githubusercontent.com/5253184/99806855-4679ec80-2b47-11eb-8a64-77003fee8c39.png)

Calling CancelViewModelAsync before closing fixes the issue- but it isn't hard to [forget/don't know] to do so, or as in my case call CancelAsync instead :)

**Comment by GeertvanHorrik** [comment]
Yes, one of the things that is required is that you call SaveAndCloseViewModel, otherwise the vm doesn't know how to clean up. One of the things we could do is throw an exception in case a VM is closed without a call to Save / Cancel.

**Comment by GeertvanHorrik** [comment]
I think throwing an exception is the best way, but not sure whether it's wise to introduce at this stage (in a hotfix release). Maybe it's better to do this for 6.0

**Comment by lipchev** [comment]
I think throwing an exception would definitely be a breaking change for some people who never noticed the memory leak (and were ok with it). A log message would seem more appropriate. 

I see your point about the cleanUpMode (it is possible that the user doesn't want to preserve the changes for some reason)- but that part of the method can be skipped, and I see no problem in cleaning up the rest (even if doing so twice).

**Comment by lipchev** [comment]
Right, so I've been profiling another application where I've got some long lived objects (cached in a single ORM context), bound to a ListView (virtualization enabled). Each item is then bound to a ChildViewModel. 
Whenever an item (ChildView) is "unloaded" (either scrolling or filtering), the ChildViewModel on OnClosed is called- I do my cleanup thinking it's all good. However due to the fact that the view model was neither canceled nor saved (in fact it hasn't even been modified yet), the memory leak is created.. Canceling the ParentViewModel afterwards does nothing for the already closed view models.

**Comment by GeertvanHorrik** [comment]
When a view is unloaded, Catel should cancel it. Are you saying it's not doing that correctly? If so I would like to see a repro.

**Comment by lipchev** [comment]
In my case the unload behavior is set to `UnloadBehavior.CloseViewModel`, I'm considering switching to `CancelAndCloseViewModel` instead, but that would need some special handling- as I'm currently using the CancelAsync method to actually cancel any changes (which is not what I want to happen when "scrolling up/down").

**Comment by lipchev** [comment]
It took a bit of work but I managed switching to using CancelAndCloseViewModel and that plugged the leaks. 


> 
> 
> When a view is unloaded, Catel should cancel it. Are you saying it's not doing that correctly? If so I would like to see a repro.

Do you still want me to make a repro of the leak using the list-of-viewmodels-that-are-not-saved-or-canceled (i.e. using the CloseViewModel unload behavior)?

**Comment by GeertvanHorrik** [comment]
Not sure. I think we can still solve this by canceling the models (at least unsubscribing from the events), but not sure whether we should throw an exception or silently unsubscribe.

1. throw exception when calling close without save / cancel first
2. allow closing without saving / canceling (goes a bit against the Catel philosophy) by unsubscribing events in the close instead of cancel / save

**Comment by lipchev** [comment]
I think the second option is the way to go- what you are describing is exactly the behavior that I would expect with the `UnloadBehavior.CloseViewModel` (given the two other alternatives). Here are a two philosophical arguments to back it up:

- I can replace the model (via the setter) at any time without calling Save/Cancel: this is arguably the same as closing and reopening a new view model (correctly unsubscribing from the previous model without any complaints)
- should I be _Saving_ or _Canceling_ a read-only model?

**Comment by GeertvanHorrik** [comment]
> should I be Saving or Canceling a read-only model?

We are not canceling models, we are canceling view models. Then it's up to the developer how to handle commands during save / cancel, etc. But I agree, we should make it "supported".

**Comment by GeertvanHorrik** [comment]
Getting back to this while finalizing Catel 6. I think we should change the default behavior to cancel and users should explicitly use SaveAndCloseViewModel if they want to save.
